pkg/transport/bridge.go

```diff
@@ -0,0 +1,404 @@
+package transport
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+
+	"golang.org/x/exp/jsonrpc2"
+
+	"github.com/stacklok/toolhive/pkg/logger"
+)
+
+// StdioBridge implements a bridge for MCP servers using stdio transport.
+type StdioBridge struct {
+	baseURL, postURL *url.URL
+	headers          http.Header
+	mode             string
+	wg               sync.WaitGroup
+	cancel           context.CancelFunc
+	InitReady        chan struct{}
+}
+
+// NewStdioBridge creates a new StdioBridge instance.
+func NewStdioBridge(rawURL string) (*StdioBridge, error) {
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return nil, fmt.Errorf("invalid URL: %w", err)
+	}
+	return &StdioBridge{baseURL: u, postURL: u}, nil
+}
+
+// Start begins the transport loop for the StdioBridge.
+func (b *StdioBridge) Start(ctx context.Context) {
+	ctx, b.cancel = context.WithCancel(ctx)
+	b.wg.Add(1)
+	go b.loop(ctx)
+}
+
+func (b *StdioBridge) loop(ctx context.Context) {
+	defer b.wg.Done()
+	for {
+		mode, err := b.detectTransport(ctx)
+		if err != nil {
+			logger.Errorf("transport detection failed: %v", err)
+			return
+		}
+		b.mode = mode
+		b.wg.Add(2)
+		if mode == "streamable-http" {
+			go b.runStreamableReader(ctx)
+			go b.runStreamableWriter(ctx)
+		} else { // legacy SSE
+			b.postURL = b.baseURL // reset
+			go b.runLegacyReader(ctx)
+			go b.runLegacyWriter(ctx)
```

Comment:
Do we need to spawn both readers in the beginning? can't we detect the transport first?

Reply:
it's reader and writer... do not understand your comment there